Throw an error if DB replication fails

johnclary opened the pull request:
## Associated issues

* https://github.com/cityofaustin/atd-data-tech/issues/22791

This update the `vision-zero` helper script to make it more obvious when a replication error occurs.



## Testing

**URL to test:** Local

In the root of this repo, modify your `.env` file by changing the RR database from `vision_zero` to `vision_zero_x`.

```shell
# .env
RR_DATABASE="vision_zero_x"
```

Now try to replicate

```shell
 ./vision-zero replicate-db
```

And confirm that the scrip throws an error. Something like:

```shell
❌ pg_dump failed with return code 1
Error output: None
```

Locate and open the failed snapshot in the `/database/snapshots` directory. It will have the `.error` file extension.

Fix your `.env` so that it's using the correct DB name, and replicate again.

While the snapshot is downloading, browser your `/database/snapshots` directory and observe that the snapshot filename has the extension `.in_progress`.

Once the replication completes, observe the file extension is `.sql` and that the database is working normally.

Finally, test out a few of the other command line options. Definitely test this one:

```shell
./vision-zero local-stack-up
```

### Bonus points

Remove or rename today's working snapshot from your `/database/snapshots` directory. Start a fresh snapshot download with ` ./vision-zero replicate-db`. Then interrupt the process with your keyboard by pressing `ctrl` + `c`. Verify that the incomplete snapshot has the extension `.error`.

---

#### Ship list

- [ ] Check migrations for any conflicts with latest migrations in `main` branch
- [ ] Confirm Hasura role permissions for necessary access
- [ ] Code reviewed
- [ ] Product manager approved

johnclary commented on vision-zero:
```diff
@@ -187,9 +175,26 @@ def replicateDb(args):
+                stdout=snapshot,
+            )
+        except Exception as e:
+            os.remove(snapshotFQP)  # Remove the incomplete file
```
removing the file prevents us from entering the doom loop of `replicate-db` sourcing the invalid snapshot file instead of fetching a new one, but it also prevents us from inspecting that file for error messages. i guess removing the snapshot is not idea? 🤔

johnclary commented on vision-zero:
```diff
@@ -283,9 +294,9 @@ def buildTools(args):
+def dbUp():
     print("🛢️  Starting database")
-    subprocess.run(
+    result = subprocess.run(
```
i decided not to add handling for every `subprocess.run()` call in the helper (there are a lot)—i just targeted the three places where issues seem to fail silently.

johnclary commented on vision-zero:
```diff
@@ -187,9 +176,28 @@ def replicateDb(args):
+            )
+        except Exception as e:
+            # rename the corrupted snapshot
+            os.rename(snapshotFQP, snapshotFQP_error)
```
renaming the incomplete snapshot file prevents us from entering the doom loop of `replicate-db` sourcing the invalid snapshot file instead of fetching a new one.

johnclary commented on vision-zero:
```diff
@@ -55,48 +55,48 @@ def checkDockerComposeAvailability():
 
 def doCommand(args):
     if args.command == "build":
-        buildTools(args)
```
i removed `args` from a number of functions which didn't use them

chiaberry commented on vision-zero:
```diff
@@ -55,48 +55,48 @@ def checkDockerComposeAvailability():
     elif args.command == "local-stack-down":
-        cr3UserApiDown(args)
+        cr3UserApiDown()
         graphqlEngineDown(args)
```
also `dbDown` calls `graphqlEngineDown` as it's first step, we could remove it from line 65 here

roseeichelmann reviewed (approved):
thank you for these additions!

i got that conflict with recovery error on first replicate, but was successful on the second try

johnclary commented:
I have re-requested review after adding a `.in_progress` file extension to the mix. I also cleaned up the args Chia pointed out. I've updated the test instructions accordingly 🙏

chiaberry reviewed (approved):
tested again, saw the .error file, .in_progress and then turned off my wifi to see how it handled it, and it changed its file extension to .error. 

Ran again without interfering and got the working db up and running. 🚢

mddilley reviewed (approved):
I followed the test steps, and I see all the expected file names and console output when seeing replication in progress, complete, or errored. 🚀 ✨
